MoreLinq/Memoize.cs

```diff
@@ -0,0 +1,136 @@
+        /// in-memory source.
+        /// An additional argument specifies if buffering must happen even if the source implements
+        /// <see cref="ICollection{T}"/> (otherwise it is assumed that the source already serves as 
+        /// a buffer adequately).
+        /// </summary>
+        /// <param name="source">The source sequence.</param>
+        /// <param name="forceBuffering">Force buffering, even if source is an <see cref="ICollection{T}" />
+        /// (otherwise source is assumed to already be in-memory and enumerated quickly, and hence that
+        /// buffering again would be a waste).</param>
+        /// <returns>A <see cref="MemoizedEnumerable{T}"/>.</returns>
+        public static IBufferedEnumerable<T> Memoize<T>(this IEnumerable<T> source, bool forceBuffering)
+        {
+            if (source == null) throw new ArgumentNullException("source");
+
+            return (source as MemoizedEnumerable<T>) ?? new MemoizedEnumerable<T>(source, forceBuffering);
+        }
+    }
+
+    /// <summary>
+    /// An enumeration that, when being iterated, lazily iterates through the source enumeration 
+    /// and cache the elements already iterated for future iterations. It supports partial iterations.
+    /// </summary>
+    /// <typeparam name="T">Type of the source sequence</typeparam>
+    public interface IBufferedEnumerable<T> : IEnumerable<T>
+    {
+
+    }
+
+    internal class MemoizedEnumerable<T> : IBufferedEnumerable<T>, IEnumerable<T>
+    {
+        private ICollection<T> collection;
+        private IEnumerable<T> source;
+        private IEnumerator<T> e;
+        private IList<T> cache;
+        private bool disposed;
+
+        public MemoizedEnumerable(IEnumerable<T> sequence, bool forceBuffering)
+        {
+            if (sequence == null) throw new ArgumentNullException("sequence");
+		
+            if (!forceBuffering && sequence is ICollection<T>)
+            {
+                collection = (ICollection<T>)sequence;
+            }
+            else
+            {
+                source = sequence;
+                cache = new List<T>();
+            }
+        }
+
+        private IEnumerator<T> GetMemoizedEnumerator()
+        {
+            if (e == null && !disposed) e = source.GetEnumerator();
+
+            int index = 0;
+
+            while (true)
+            {
+                while (index < cache.Count)
+                {
+                    yield return cache[index++];
+                }
+
+                if (!disposed && e.MoveNext())
+                {
+                    cache.Add(e.Current);
+                    index++;
+                    yield return e.Current;
+                }
+                else
+                {
+                    if (!disposed)
+                    {
+                        disposed = true;
+                        e.Dispose();
+                        source = null;
+                        e = null;
+                    }
```

Comment:
>  would make it impossible to chain the operator

You should be able to chain via [`Using`](https://github.com/Reactive-Extensions/Rx.NET/blob/v3.1.1/Ix.NET/Source/System.Interactive/Using.cs) from [System.Interactive](https://www.nuget.org/packages/System.Interactive/).

I'd rather do away with the Boolean in a first cut, while maintaining the functionality via `IDisposable` where it may be critically important, and bring it back (possibly taking another form) if real-world uses demonstrate that it was a good idea in the first place. However, taking it out later is always more painful.

Disposal would also add the idea of flushing the internal cache (think refresh). It wouldn't be so much a disposable of the enumerable itself than the internal state (perhaps an abuse of the idea). As the implementation currently stands, the cache will live as long as you hold on to the enumerable (think static field) because [it's a read-only field](https://github.com/morelinq/MoreLINQ/pull/211/files#diff-0cc36bebb7111c5eb462ec0ad60bdd27R87).